repo/object/object_reader.go

```diff
@@ -102,7 +106,38 @@ func (r *objectReader) Read(buffer []byte) (int, error) {
 	return readBytes, nil
 }
 
+func (r *objectReader) readAhead(ctx context.Context) {
+	if r.opts.ReadAheadBytes <= 0 {
+		return
+	}
+
+	var (
+		contentIDs  []content.ID
+		totalLength int64
+	)
+
+	for i := r.currentChunkIndex + 1; i < len(r.seekTable); i++ {
+		if cid, _, ok := r.seekTable[i].Object.ContentID(); ok {
+			if i > r.maxReadAheadChunkIndex {
+				contentIDs = append(contentIDs, cid)
+				r.maxReadAheadChunkIndex = i
+			}
+
+			totalLength += r.seekTable[i].Length
+			if totalLength > r.opts.ReadAheadBytes {
+				break
+			}
+		}
+	}
+
+	if len(contentIDs) > 0 {
+		go r.cr.PrefetchContents(ctx, contentIDs, "blobs")
+	}
+}
+
 func (r *objectReader) openCurrentChunk() error {
+	r.readAhead(r.ctx)
```

Comment:
One down side of this approach is that we won't pull the next content until we read the first byte. Also, if you are reading large content blobs then you are constantly hitting your prefetch boundary on each `openCurrentChunk()`, hence this will not prefetch until read if avg object size is close to `ReadAheadBytes`.

How about starting long running prefetch workers at the time of creation of the object-reader ? That way you are not continuously jump-starting your prefetch routines every time you hit a boundary.

Reply:
The time between reading first byte and opening the reader is usually negligible. The upside of not reading immediately is that it lets you check the length and seek in the reader before you start.